src/Framework/AspNetCoreAnalyzers/src/Analyzers/Mvc/MvcAnalyzer.cs

```diff
@@ -0,0 +1,152 @@
+                }
+            }, SymbolKind.NamedType);
+        });
+    }
+
+    private static void PopulateActionRoutes(SymbolAnalysisContext context, WellKnownTypes wellKnownTypes, RouteUsageCache routeUsageCache, INamedTypeSymbol namedTypeSymbol, List<ActionRoute> actionRoutes)
+    {
+        foreach (var member in namedTypeSymbol.GetMembers())
+        {
+            if (member is IMethodSymbol methodSymbol &&
+                MvcDetector.IsAction(methodSymbol, wellKnownTypes))
+            {
+                foreach (var attribute in methodSymbol.GetAttributes())
+                {
+                    if (attribute.AttributeClass is null || !wellKnownTypes.IsType(attribute.AttributeClass, RouteAttributeTypes, out var match))
+                    {
+                        continue;
+                    }
+
+                    var routeUsage = GetRouteUsageModel(attribute, routeUsageCache, context.CancellationToken);
+                    if (routeUsage is null)
+                    {
+                        continue;
+                    }
+
+                    var httpMethodsBuilder = ImmutableArray.CreateBuilder<string>();
+
+                    switch (match.Value)
+                    {
+                        case WellKnownType.Microsoft_AspNetCore_Mvc_RouteAttribute:
+                            // No HTTP method.
+                            break;
+                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpDeleteAttribute:
+                            httpMethodsBuilder.Add("DELETE");
+                            break;
+                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpGetAttribute:
+                            httpMethodsBuilder.Add("GET");
+                            break;
+                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpHeadAttribute:
+                            httpMethodsBuilder.Add("HEAD");
+                            break;
+                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpOptionsAttribute:
+                            httpMethodsBuilder.Add("OPTIONS");
+                            break;
+                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpPatchAttribute:
+                            httpMethodsBuilder.Add("PATCH");
+                            break;
+                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpPostAttribute:
+                            httpMethodsBuilder.Add("POST");
+                            break;
+                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpPutAttribute:
+                            httpMethodsBuilder.Add("PUT");
+                            break;
+                        default:
+                            throw new InvalidOperationException("Unexpected well known type:" + match);
+                    }
+                    actionRoutes.Add(new ActionRoute(methodSymbol, routeUsage, httpMethodsBuilder.ToImmutable()));
```

Comment:
No need to allocate a builder, I think.

It might not even be necessary to have an immutable array. Could a simple string work where `null` means no HTTP method (the empty array case)

```suggestion
                    var httpMethods = ImmutableArray.Empty<string>();

                    switch (match.Value)
                    {
                        case WellKnownType.Microsoft_AspNetCore_Mvc_RouteAttribute:
                            // No HTTP method.
                            break;
                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpDeleteAttribute:
                            httpMethods = ImmutableArray.Create("DELETE");
                            break;
                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpGetAttribute:
                            httpMethods = ImmutableArray.Create("GET");
                            break;
                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpHeadAttribute:
                            httpMethods = ImmutableArray.Create("HEAD");
                            break;
                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpOptionsAttribute:
                            httpMethods = ImmutableArray.Create("OPTIONS");
                            break;
                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpPatchAttribute:
                            httpMethods = ImmutableArray.Create("PATCH");
                            break;
                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpPostAttribute:
                            httpMethods = ImmutableArray.Create("POST");
                            break;
                        case WellKnownType.Microsoft_AspNetCore_Mvc_HttpPutAttribute:
                            httpMethods = ImmutableArray.Create("PUT");
                            break;
                        default:
                            throw new InvalidOperationException("Unexpected well known type:" + match);
                    }
                    actionRoutes.Add(new ActionRoute(methodSymbol, routeUsage, httpMethods));
```